### PR TITLE
Enable `CopyHtmlAttribute` to maintain copied attribute order.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Updated implementation to do a best guess at copying an attribute back into `TagHelperOutput.Attributes`.
- Updated existing functional and unit tests to account for maintained attribute order.
- Added a set of complex order based unit tests to validate `CopyHtmlAttribute` properly maintains order.

#2639

### CODE REVIEW

**Comment by NTaylorMullen** [comment]
Making a change to the algorithm. Instead of going backwards then forwards through the `context.AllAttributes` to find a familiar attribute i'm going to look both ways simultaneously to ensure we find the closest.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs [hunk 1]
```diff
@@ -8,13 +8,258 @@
...
+                            new TagHelperAttribute("secondDynamic", "another value"),
+                        }
+                    },
+                    {
+                        "first",
+                        new TagHelperAttributeList
+                        {
+                            { "third", "C" },
+                            { "dynamic", "value" },
+                            { "secondDynamic", "another value"}
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: How would `output.Attributes` get into this mixed-up state?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs [hunk 2]
```diff
@@ -8,13 +8,258 @@
...
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: Again, is this mix-up realistic? Does behaviour once in this state matter much?

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 3]
```diff
@@ -33,21 +33,27 @@ public static class TagHelperOutputExtensions
...
+                for (var i = context.AllAttributes.Count - 1; i >= 0; i--)
+                {
+                    // We look for the original attribute so we can restore the exact attribute name the user typed.
+                    // Approach also ignores changes made to tagHelperOutput[attributeName].
+                    if (string.Equals(
+                        attributeName,
+                        context.AllAttributes[i].Name,
+                        StringComparison.OrdinalIgnoreCase))
+                    {
+                        CopyHtmlAttributeMaintainOrder(i, tagHelperOutput, context);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Suggest calling the called method `CopyHtmlAttribute()`. Since both the current method and the one called here maintain order, that's not a distinguishing feature.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 4]
```diff
@@ -33,21 +33,27 @@ public static class TagHelperOutputExtensions
         {
             if (!tagHelperOutput.Attributes.ContainsName(attributeName))
             {
-                IEnumerable<IReadOnlyTagHelperAttribute> entries;
+                var copiedAttribute = false;
+                for (var i = context.AllAttributes.Count - 1; i >= 0; i--)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Add a comment about why this needs to go through `AllAttributes` backwards.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 5]
```diff
@@ -33,21 +33,27 @@ public static class TagHelperOutputExtensions
...
+                {
+                    // We look for the original attribute so we can restore the exact attribute name the user typed.
+                    // Approach also ignores changes made to tagHelperOutput[attributeName].
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Update comment to mention restoring the attribute (approximately) where the user wrote it in the Razor source.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 6]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
                 tagHelperOutput.Attributes.Remove(attribute);
             }
         }
+
+        private static void CopyHtmlAttributeMaintainOrder(
+            int allAttributeIndex,
+            TagHelperOutput tagHelperOutput,
+            TagHelperContext context)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Probably worth sticking with the `TagHelperOutput`, {specific attribute info}, `TagHelperContext` order of the (other) `CopyHtmlAttribute()` method.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 7]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+            {
+                Name = existingAttribute.Name,
+                Value = existingAttribute.Value,
+                Minimized = existingAttribute.Minimized
+            };
+
+            for (var i = allAttributeIndex - 1; i >= 0; i--)
+            {
+                // Move backwards through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value in front of.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Move comment above the loop since it doesn't just describe the loop body.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 8]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+                    attributes: tagHelperOutput.Attributes))
+                {
+                    return;
+                }
+            }
+
+            for (var i = allAttributeIndex + 1; i < context.AllAttributes.Count; i++)
+            {
+                // Move forward through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Also move this comment above the loop.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 9]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+                Name = existingAttribute.Name,
+                Value = existingAttribute.Value,
+                Minimized = existingAttribute.Minimized
+            };
+
+            for (var i = allAttributeIndex - 1; i >= 0; i--)
+            {
+                // Move backwards through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value in front of.
+                var previousAllAttributeName = context.AllAttributes[i].Name;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Bit of a mouthful; suggest `previousName`.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 10]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+                {
+                    return;
+                }
+            }
+
+            for (var i = allAttributeIndex + 1; i < context.AllAttributes.Count; i++)
+            {
+                // Move forward through context.AllAttributes from the provided index until we find a familiar
+                // attribute in tagHelperOutput where we can insert the copied value.
+                var nextAllAttributeName = context.AllAttributes[i].Name;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Suggest `nextName`.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 11]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+            tagHelperOutput.Attributes.Add(copiedAttribute);
+        }
+
+        private static bool TryFindNameAndInsertAttribute(
+            string name,
+            int insertOffset,
+            TagHelperAttribute attributeToInsert,
+            TagHelperAttributeList attributes)
+        {
+            for (var j = 0; j < attributes.Count; j++)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Use `i` or `index`. Neither is ambiguous here.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 12]
```diff
@@ -100,5 +106,70 @@ public static class TagHelperOutputExtensions
...
+            {
+                if (string.Equals(name, attributes[j].Name, StringComparison.OrdinalIgnoreCase))
+                {
+                    attributes.Insert(j + insertOffset, attributeToInsert);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: This results in an odd mix of responsibilities. Instead do the insertion in `CopyHtmlAttribute()` and change this method to
``` c#
private static int FindFirstMatch(TagHelperAttributeList attributeList, string name)
{
    for (var index = 0; index < attributes.Count; index++)
    {
        if (string.Equals(name, attributeList[index].Name, StringComparison.OrdinalIgnoreCase))
        {
            return index;
        }
    }

    return -1;
}
```

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs (hunk 6 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Nah, other `CopyHtmlAttribute` method is an extension. Otherwise it wouldn't have same order.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 1 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]:
```C#
output.CopyHtmlAttribute("third", context);
...
output.Attributes.Add("dynamic", "value");
output.Attributes.Add("secondDynamic", "another value");
```

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: We modify attributes somewhat similarly. Same setup could occur: https://github.com/aspnet/Mvc/pull/2987/files#r37362253

**Comment by NTaylorMullen** [comment]
Updated.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 1 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: :ok:

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: :ok:

**Comment by NTaylorMullen** [comment]
This good?

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: Suggest adding another test case with `outputAttributes` containing some duplicates.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 13]
```diff
@@ -33,20 +33,30 @@ public static class TagHelperOutputExtensions
...
+                // We iterate context.AllAttributes backwards since we prioritize TagHelperOutput values occurring
+                // before the current context.AllAttribtes[i].
+                for (var i = context.AllAttributes.Count - 1; i >= 0; i--)
                 {
-                    throw new ArgumentException(
-                        Resources.FormatTagHelperOutput_AttributeDoesNotExist(attributeName, nameof(TagHelperContext)),
-                        nameof(attributeName));
+                    // We look for the original attribute so we can restore the exact attribute name the user typed in
+                    // approximately the same position where the user wrote it in the Razor source.
+                    // Approach also ignores changes made to tagHelperOutput[attributeName].
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: What `changes made to tagHelperOutput[attributeName]`? Have already confirmed this `attributeName` does not appear in the output collection.

**Comment by ajaybhargavb** [comment]
:shipit: from me

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 14]
```diff
@@ -100,5 +110,61 @@ public static class TagHelperOutputExtensions
...
+                Name = existingAttribute.Name,
+                Value = existingAttribute.Value,
+                Minimized = existingAttribute.Minimized
+            };
+
+            // Move backwards through context.AllAttributes from the provided index until we find a familiar attribute
+            // in tagHelperOutput where we can insert the copied value in front of.
+            for (var i = allAttributeIndex - 1; i >= 0; i--)
+            {
+                var previouName = context.AllAttributes[i].Name;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: tyo: `previousName`

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: My keyboard doesn't like `s`s after u's for whatever reason...

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 15]
```diff
@@ -100,5 +110,61 @@ public static class TagHelperOutputExtensions
...
+                var index = IndexOfFirstMatch(previouName, tagHelperOutput.Attributes);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Don't we want the _last_ attribute with same `Name` here? For example,
``` html
  <element existing="value 1" existing="value 2" toBeCopied="value 3"/>
```
will currently be restored as
``` html
  <element existing="value 1" toBeCopied="value 3" existing="value 2"/>
```

Do tests cover this case i.e. `output.Attributes` contains duplicate attributes one of which should be placed just before the copied attribute?

Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Suspect this code needs to figure out how many instances of `previousName` are in `context.AllAttributes[0..allAttributeIndex -1]` and look forward through `output.Attributes` to find the same count (or the last match).

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: +1

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 16]
```diff
@@ -100,5 +110,61 @@ public static class TagHelperOutputExtensions
...
+                    return;
+                }
+            }
+
+            // Move forward through context.AllAttributes from the provided index until we find a familiar attribute in
+            // tagHelperOutput where we can insert the copied value.
+            for (var i = allAttributeIndex + 1; i < context.AllAttributes.Count; i++)
+            {
+                var nextName = context.AllAttributes[i].Name;
+                var index = IndexOfFirstMatch(nextName, tagHelperOutput.Attributes);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Same problem with duplicate matches for `nextName` here.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: Plus cases where `attributeNameToCopy` matches adjacent attributes in `allAttributes`.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs (hunk 13 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Old comment. Will delete.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs (hunk 15 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Talked in person.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs (hunk 16 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: Ditto.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs (hunk 2 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/TagHelperOutputExtensionsTest.cs]: Done.

**Comment by NTaylorMullen** [comment]
Updated.

**Comment by NTaylorMullen** [comment]
Updated remarks.

**Comment by dougbu** [comment]
:shipit:
